lib/likelihood.c

```diff
@@ -0,0 +1,304 @@
+        *lik = -DBL_MAX;
+    } else {
+        *lik = 0;
+    }
+out:
+    if (tree != NULL) {
+        free(tree);
+    }
+    if (edge != NULL) {
+        free(edge);
+    }
+    if (parent != NULL) {
+        free(parent);
+    }
+    if (child != NULL) {
+        free(child);
+    }
+    return ret;
+}
+
+int
+msp_log_likelihood_arg(tsk_treeseq_t *ts, double rho, double *lik) {
+    int ret = 1;
+    int i;
+    int lineages = (int)tsk_treeseq_get_num_samples(ts);
+    tsk_id_t parent_id;
+    double sim_time = 0;
+    double material = lineages * tsk_treeseq_get_sequence_length(ts);
+    double material_in_children;
+    double material_in_parent;
+    double rate;
+    double gap;
+    int first_parent_edge[ts->tables->nodes.num_rows];
+    int last_parent_edge[ts->tables->nodes.num_rows];
+    tsk_edge_t *edge = NULL;
+    tsk_node_t *parent = NULL;
+    tsk_node_t *child = NULL;
+    edge = malloc(sizeof *edge);
+    if (edge == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    parent = malloc(sizeof *parent);
+    if (parent == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    child = malloc(sizeof *child);
+    if (child == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    for (i = 0; i < (int)ts->tables->nodes.num_rows; i++) {
+        first_parent_edge[i] = -1;
+        last_parent_edge[i] = -1;
+    }
+    for (i = 0; i < (int)ts->tables->edges.num_rows; i++) {
+        ret = tsk_treeseq_get_edge(ts, i, edge);
+        if (ret != 0) {
+            goto out;
+        }
+        if (first_parent_edge[edge->child] == -1) {
+            first_parent_edge[edge->child] = i;
+        }
+        last_parent_edge[edge->child] = i;        
+    }
+    ret = tsk_treeseq_get_edge(ts, 0, edge);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->child, child);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->parent, parent);
+    if (ret != 0) {
+        goto out;
+    }
+    *lik = 0;
+    while (lineages > 0) {
+        rate = lineages * (lineages - 1) / 2 + material * rho;
+        *lik -= rate * (parent->time - sim_time);
+        sim_time = parent->time;
+        parent_id = parent->id;
+        if (parent->flags == MSP_NODE_IS_RE_EVENT) {
+            while (msp_increment_edge(ts, edge, parent, child) == 0 &&
+                    parent->id == parent_id) {
+                continue;
+            }
+            gap = edge->left - ts->tables->edges.right[edge->id - 1];
+            ret = tsk_treeseq_get_edge(ts, last_parent_edge[edge->child], edge);
+            if (ret != 0) {
+                goto out;
+            }
+            material -= gap;
+            lineages++;
+            if (!(gap > 0)) {
+                // we evaluate the density rather than probability
+                gap = 1;
+            }
+            if (!(rho > 0)) {
+                *lik -= DBL_MAX;
```

Comment:
Why -=  DBL_MAX? Wouldn't you be setting it to DBL_MAX either way?

Reply:
You're right. That should be *lik = -DBL_MAX;